gluten-core/src/main/scala/org/apache/gluten/extension/columnar/OffloadSingleNode.scala

```diff
@@ -181,7 +185,127 @@ case class OffloadJoin() extends OffloadSingleNode with LogLevelUtil {
+      case _: InputFileBlockStart => true
+      case _: InputFileBlockLength => true
+      case _ => expr.children.exists(isInputFileNameExpr)
+    }
+  }
+
+  private def rewriteExpr(
+      expr: Expression,
+      replacedExprs: Map[String, AttributeReference]): Expression = {
+    expr match {
+      case _: InputFileName =>
+        replacedExprs.getOrElseUpdate(
+          expr.prettyName,
+          AttributeReference(expr.prettyName, StringType, false)())
+      case _: InputFileBlockStart =>
+        replacedExprs.getOrElseUpdate(
+          expr.prettyName,
+          AttributeReference(expr.prettyName, LongType, false)())
+      case _: InputFileBlockLength =>
+        replacedExprs.getOrElseUpdate(
+          expr.prettyName,
+          AttributeReference(expr.prettyName, LongType, false)())
+      case other =>
+        other.withNewChildren(other.children.map(child => rewriteExpr(child, replacedExprs)))
+    }
+  }
 
+  private def addMetadataCol(
+      plan: SparkPlan,
+      replacedExprs: Map[String, AttributeReference]): SparkPlan = {
+    def genNewOutput(output: Seq[Attribute]): Seq[Attribute] = {
+      var newOutput = output
+      for ((_, newAttr) <- replacedExprs) {
+        if (!newOutput.exists(attr => attr.exprId == newAttr.exprId)) {
+          newOutput = newOutput :+ newAttr
+        }
+      }
+      newOutput
+    }
+    def genNewProjectList(projectList: Seq[NamedExpression]): Seq[NamedExpression] = {
+      var newProjectList = projectList
+      for ((_, newAttr) <- replacedExprs) {
+        if (!newProjectList.exists(attr => attr.exprId == newAttr.exprId)) {
+          newProjectList = newProjectList :+ newAttr.toAttribute
+        }
+      }
+      newProjectList
+    }
+    plan match {
+      case f: FileSourceScanExec =>
+        f.copy(output = genNewOutput(f.output))
+      case f: FileSourceScanExecTransformer =>
+        f.copy(output = genNewOutput(f.output))
+      case b: BatchScanExec =>
+        b.copy(output = genNewOutput(b.output).asInstanceOf[Seq[AttributeReference]])
+      case b: BatchScanExecTransformer =>
+        b.copy(output = genNewOutput(b.output).asInstanceOf[Seq[AttributeReference]])
+      case p @ ProjectExec(projectList, child) =>
+        p.copy(genNewProjectList(projectList), addMetadataCol(child, replacedExprs))
+      case p @ ProjectExecTransformer(projectList, child) =>
+        p.copy(genNewProjectList(projectList), addMetadataCol(child, replacedExprs))
+      case _ => plan.withNewChildren(plan.children.map(addMetadataCol(_, replacedExprs)))
+    }
+  }
+
+  private def genProjectExec(projectExec: ProjectExec): SparkPlan = {
+    def findScanNodes(plan: SparkPlan): Seq[SparkPlan] = {
+      plan match {
+        case f @ (_: FileSourceScanExec | _: AbstractFileSourceScanExec |
+            _: DataSourceV2ScanExecBase) =>
+          Seq(f)
+        case o => o.children.flatMap(findScanNodes)
+      }
+    }
+
+    if (!TransformHints.isNotTransformable(projectExec)) {
```

Comment:
!TransformHints.isNotTransformable -> TransformHints.isTransformable, easier to understand

Reply:
Although gluten code seems warn that isTransformable is not reliable:

```
  /**
   * NOTE: To be deprecated. Do not create new usages of this method.
   *
   * Since it's usually not safe to consider a plan "transformable" during validation phase. Another
   * validation rule could turn "transformable" to "non-transformable" before implementing the plan
   * within Gluten transformers.
   */
  def isTransformable(plan: SparkPlan): Boolean = {
```

Which also mean `!TransformHints.isNotTransformable` is not reliable, right? not sure.

Reply:
`TransformHints.isNotTransformable` shoud re reliable, we shouldn't use `TransformHints.isTransformable`

Reply:
According to the doc of `isNotTransformable`, if `isNotTransformable == false`, the result is not reliable. Actually
`!TransformHints.isNotTransformable == TransformHints.isTransformable`, if `TransformHints.isTransformable` is not reliable, `!TransformHints.isNotTransformable` is not reliable either, right? Any insights? @zhztheplayer 

```
  /**
   * If true, the plan node will be guaranteed fallback to Vanilla plan node while being
   * implemented.
   *
   * If false, the plan still has chance to be turned into "non-transformable" in any another
   * validation rule. So user should not consider the plan "transformable" unless all validation
   * rules are passed.
   */
  def isNotTransformable(plan: SparkPlan): Boolean = {
```

Reply:
@gaoyangxiaozhu This comment has not been resolved yet.